lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp

```diff
@@ -318,24 +318,15 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
              binder.tensorOperands(operands, 9)) ||
             binder.tensorResultType(resultType))
           return failure();
-        Value a = operands[0];
-        Value aScale = operands[1];
-        Value aZp = operands[2];
-        Value b = operands[3];
-        Value bScale = operands[4];
-        Value bZp = operands[5];
-        Value cScale = operands[6];
-        Value cZp = operands[7];
-        Value c = operands.size() == 9 ? operands[8] : nullptr;
-
-        auto check = [](Value v) {
-          auto vTy = cast<Torch::ValueTensorType>(v.getType());
-          return llvm::all_of(vTy.getSizes(), [](int64_t d) { return d == 1; });
-        };
-        if (!check(aScale) || !check(aZp) || !check(bScale) || !check(bZp) ||
-            !check(cScale) || !check(cScale))
-          return rewriter.notifyMatchFailure(
-              binder.op, "not supported for non per-tensor quantization");
+        Value input = operands[0];
+        Value inputScale = operands[1];
+        Value inputZp = operands[2];
+        Value weight = operands[3];
+        Value weightScale = operands[4];
+        Value weightZp = operands[5];
+        Value outputScale = operands[6];
+        Value outputZp = operands[7];
+        Value output = operands.size() == 9 ? operands[8] : nullptr;
```

Comment:
@vivekkhandelwal1 this can be easily wrapped up in a helper function and then unpacked using structured bindings.

The following diff should work:

```diff
diff --git a/include/torch-mlir/Conversion/Utils/Utils.h b/include/torch-mlir/Conversion/Utils/Utils.h
index 264fb496..4a00c324 100644
--- a/include/torch-mlir/Conversion/Utils/Utils.h
+++ b/include/torch-mlir/Conversion/Utils/Utils.h
@@ -18,6 +18,21 @@ namespace mlir {
 namespace torch {
 namespace Torch {
 
+template <typename ElemType, int sz>
+std::optional<std::array<ElemType, sz>>
+takeFirstN(SmallVector<ElemType> &container) {
+  if (sz > container.size()) {
+    return std::nullopt;
+  }
+
+  std::array<ElemType, sz> firstNElems;
+  for (size_t i = 0; i < sz; ++i) {
+    firstNElems[i] = container[i];
+  }
+
+  return firstNElems;
+}
+
 LogicalResult verifyLinalgCompatibleTypes(Operation *op,
                                           PatternRewriter &rewriter);
 
diff --git a/lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp b/lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp
index 2a6f42a4..3f8361b1 100644
--- a/lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp
+++ b/lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp
@@ -9,6 +9,7 @@
 
 #include "torch-mlir/Conversion/TorchOnnxToTorch/Patterns.h"
 #include "torch-mlir/Conversion/TorchOnnxToTorch/Utils.h"
+#include "torch-mlir/Conversion/Utils/Utils.h"
 #include "torch-mlir/Dialect/Torch/Utils/Utils.h"
 #include "llvm/ADT/ArrayRef.h"
 #include "llvm/ADT/SmallVector.h"
@@ -318,14 +319,11 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
              binder.tensorOperands(operands, 9)) ||
             binder.tensorResultType(resultType))
           return failure();
-        Value input = operands[0];
-        Value inputScale = operands[1];
-        Value inputZp = operands[2];
-        Value weight = operands[3];
-        Value weightScale = operands[4];
-        Value weightZp = operands[5];
-        Value outputScale = operands[6];
-        Value outputZp = operands[7];
+
+        const auto firstEightOperands = Torch::takeFirstN<Value, 8>(operands);
+        auto [input, inputScale, inputZp, weight, weightScale, weightZp,
+              outputScale, outputZp] = firstEightOperands.value();
+
         Value output = operands.size() == 9 ? operands[8] : nullptr;
 
         auto extract = [&rewriter, &binder](Value v) {
```

Passing the size as a template parameter makes it known at compile-time, otherwise the unpacking would not be allowed by the compiler.

Reply:
Okay. There might be many ways of doing this but what are we trying to achieve via this change? Instead, it increases  the code size and I don't see it more optimal than the former one in any way.

Reply:
Well, IMO splitting this logic into a function is useful as this makes it reusable. I can't see how exactly it would make the binary any slower. Writing 10 lines of code every time you need the first 10 elements of an array appears to me to warrant a function, but, of course, YMMV.

Reply:
Yeah, it's definitely good to have something in reusable form but that's beneficial when there are multiple uses of it. I can't find multiple uses of this except 1 (QLinearMatmul). And, I think the original author wrote it that way to make it more readable otherwise all of that is nothing more than just a single line of code. 

Anyway, sorry for the disagreement but I don't find moving this to a utility more helpful. So, I would like to keep it like that unless otherwise we have more people supporting that argument.

Reply:
No, that is alright, of course. We can leave this, unless of course other reviewers think otherwise.